classes/local/store/azure_blob_storage/client.php

```diff
@@ -0,0 +1,296 @@
+            return false;
+        }
+    }
+
+    /**
+     * Returns a stream context used to handle file IO
+     * @return resource stream resource
+     */
+    public function get_seekable_stream_context() {
+        $context = stream_context_create([
+            'blob' => [
+                'seekable' => true,
+            ],
+        ]);
+        return $context;
+    }
+
+    /**
+     * Test permissions by uploading and doing various actions.
+     * @param bool $testdelete if should test deletion.
+     * @return stdClass containing 'success' and 'messages' values.
+     */
+    public function test_permissions($testdelete): stdClass {
+        $key = 'permissions_check_test';
+        $file = Utils::streamFor('test permission file');
+        $filemd5 = hex2bin(md5('test permission file'));
+
+        // Try create a file.
+        try {
+            $this->api->put_blob_async($key, $file, $filemd5)->wait();
+        } catch (Throwable $e) {
+            return (object) [
+                'success' => false,
+                'messages' => [get_string('settings:writefailure', 'tool_objectfs') . $e->getMessage() => 'notifyproblem'],
+            ];
+        }
+
+        // Try read the file that was created.
+        try {
+            $this->api->get_blob_async($key, $file, $filemd5)->wait();
+        } catch (Throwable $e) {
+            return (object) [
+                'success' => false,
+                'messages' => [get_string('settings:permissionreadfailure', 'tool_objectfs') . $e->getMessage() => 'notifyproblem'],
+            ];
+        }
+
+        // If testing delete, try delete the test file.
+        if ($testdelete) {
+            try {
+                $this->api->delete_blob_async($key)->wait();
+            } catch (Throwable $e) {
+                return (object) [
+                    'success' => false,
+                    'messages' => [get_string('settings:deleteerror', 'tool_objectfs') . $e->getMessage() => 'notifyproblem'],
+                ];
+            }
+        }
+
+        return (object) [
+            'success' => true,
+            'messages' => [get_string('settings:permissioncheckpassed', 'tool_objectfs') => 'notifysuccess'],
+        ];
+    }
+
+    /**
+     * Tests connection
+     * @return stdClass with 'success' and 'details' values.
+     */
+    public function test_connection(): stdClass {
+        // Try to create a file.
+        try {
+            $this->api->put_blob_async('connection_check_test', Utils::streamFor('test contents'), hex2bin(md5('test contents')));
+        } catch (Throwable $e) {
+            return (object) [
+                'success' => false,
+                'details' => $e->getMessage(),
+            ];
+        }
+
+        return (object) [
+            'success' => true,
+            'details' => '',
+        ];
+    }
+
+    /**
+     * Returns token expiry time
+     * @return int
+     */
+    public function get_token_expiry_time(): int {
```

Comment:
A part of me wonders if this should be in the API layer

Reply:
Done